server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/AsyncExecuteContext.java

```diff
@@ -0,0 +1,229 @@
+            hook.accept(session, traceId);
+        }
+    }
+
+    public void startQuery(CountDownLatch latch) {
+        count++;
+        if (handle != null && !handle.isDone()) {
+            handle.cancel(true);
+        }
+        if (queryRuntimeTrace) {
+            handle = executor.submit(() -> {
+                try {
+                    if (!latch.await(1100, TimeUnit.MILLISECONDS)) {
+                        String traceId = session.getSyncJdbcExecutor(BACKEND_DS_KEY)
+                                .execute((StatementCallback<String>) stmt -> OBUtils
+                                        .queryTraceIdFromASH(stmt, sessionIds, session.getConnectType()));
+                        if (traceId != null) {
+                            setCurrentTraceId(traceId);
+                        }
+                    }
+                } catch (InterruptedException e) {
+                    log.warn("Failed to get trace id.", e);
+                }
+                return null;
+            });
+        }
+    }
+
+    public void finishQuery(List<JdbcGeneralResult> results) {
+        if (handle != null && !handle.isDone()) {
+            handle.cancel(true);
+        }
+        SecurityContextUtils.setCurrentUser(user);
+        try {
+            for (JdbcGeneralResult result : results) {
+                this.results.add(mapResult(result));
+                for (BiConsumer<ConnectionSession, String> hook : traceIdHooks) {
+                    hook.accept(session, result.getTraceId());
+                }
+            }
+        } finally {
+            SecurityContextUtils.clear();
+        }
+    }
+
+    private List<String> initSessionIds(ConnectionSession session) {
+        if (!queryRuntimeTrace) {
+            return null;
+        }
+        String proxySessId = ConnectionSessionUtil.getConsoleConnectionProxySessId(session);
+        if (StringUtils.isEmpty(proxySessId)) {
+            return Collections.singletonList(ConnectionSessionUtil.getConsoleConnectionId(session));
+        }
+        return session.getSyncJdbcExecutor(CONSOLE_DS_KEY).execute((StatementCallback<List<String>>) stmt -> OBUtils
+                .querySessionIdsByProxySessId(stmt, proxySessId, session.getConnectType()));
+    }
+
+    private SqlExecuteResult mapResult(JdbcGeneralResult jdbcResult) {
+        SqlExecuteResult res = generateResult(session, jdbcResult, contextMap);
+        TraceWatch traceWatch = res.getSqlTuple().getSqlWatch();
+        try (TraceStage stage = traceWatch.start(SqlExecuteStages.SQL_AFTER_CHECK)) {
+            sqlInterceptService.afterCompletion(res, session, contextMap);
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
+        }
+        if (!traceWatch.isClosed()) {
+            traceWatch.close();
+        }
+        return res;
+    }
+
+    private SqlExecuteResult generateResult(@NonNull ConnectionSession connectionSession,
```

Comment:
we'd better not do this on the context and this way may make context too much complex

Reply:
moved there logics into ODCStatementCallBack